### ISSUE TITLE
Bug when passing a function as parameter in a delayed function

### ISSUE THREAD

**Issue by TomDLT** [joblib/joblib#643]
The following script was working on `0.10.0`:
```py
from joblib import Parallel, delayed

def twice(a):
    return 2 * a

def run(method, a):
    return method(a)

if __name__ == '__main__':
    methods = [twice, ]

    results = Parallel(n_jobs=2)(delayed(run)(
        method=method, a=10) for method in methods)
```
Now it returns : 
```pytb
Traceback (most recent call last):
  File "/cal/homes/tdupre/work/src/joblib/joblib/externals/loky/process_executor.py", line 338, in _process_worker
    call_item = call_queue.get(block=True, timeout=timeout)
  File "/cal/homes/tdupre/miniconda3/envs/py36/lib/python3.6/multiprocessing/queues.py", line 113, in get
    return _ForkingPickler.loads(res)
AttributeError: Can't get attribute 'twice' on <module 'joblib.externals.loky.backend.popen_loky_posix' from '/cal/homes/tdupre/work/src/joblib/joblib/externals/loky/backend/popen_loky_posix.py'>
Traceback (most recent call last):
  File "test.py", line 17, in <module>
    method=method, a=10) for method in methods)
  File "/cal/homes/tdupre/work/src/joblib/joblib/parallel.py", line 814, in __call__
    self.retrieve()
  File "/cal/homes/tdupre/work/src/joblib/joblib/parallel.py", line 716, in retrieve
    self._output.extend(job.get(timeout=self.timeout))
  File "/cal/homes/tdupre/work/src/joblib/joblib/_parallel_backends.py", line 402, in wrap_future_result
    return future.result(timeout=timeout)
  File "/cal/homes/tdupre/work/src/joblib/joblib/externals/loky/_base.py", line 431, in result
    return self.__get_result()
  File "/cal/homes/tdupre/work/src/joblib/joblib/externals/loky/_base.py", line 382, in __get_result
    raise self._exception
joblib.externals.loky.process_executor.BrokenExecutor: A process in the process pool was terminated abruptly while the future was running or pending.
```
A git bisect leads to #516

**Comment by lesteve** [comment]
I can reproduce this one. Thanks for the bug report! I will let the loky expert speculate on this one.

**Comment by ogrisel** [comment]
This script would have not worked with joblib 0.10 under windows though.

The problem is that we use cloudpickle to pickle interactively defined functions (functions defined in the `__main__` module) only when wrapped by `delayed`. Using cloudpickle on the args is problematic as it can be significantly slower.

Maybe we can register a specific cloudpickle-based reducer for functions defined in `__main__` in the loky backend.

/cc @tomMoral .

**Comment by tomMoral** [comment]
The script is probably failing with unix and `0.10.0` when using the `spawn` start method in multiprocessing. (this script is only usable with `fork`)

I think using a custom reducer for function defined in `__main__` would actually improve loky as it means that we use the "slow" `cloudpickle` only for the `__main__` functions, regardless where they are used.

**Comment by j3mdamas** [comment]
Hi all,

We're also facing similar problem with joblib 0.12 on Linux, while it works on 0.11:

```python
[joao@computer02 htmd]$ python htmd/projections/metricrmsd.py 
/home/joao/maindisk/SANDBOX/acemanager/latest/lib/python3.6/site-packages/htmd/molecule/readers.py:778: FutureWarning: Method .as_matrix will be removed in a future version. Use .values instead.
  serials = parsedtopo.serial.as_matrix()
Creating simlist: 100%|███████████████████████████████████████████████████████████████████████████████████████████| 2/2 [00:00<00:00, 1795.89it/s]
2018-07-06 12:16:46,713 - htmd.molecule.readers - WARNING - Integer values detected in "insertion" field. Your resids might be overflowing. Take care
/home/joao/maindisk/SANDBOX/acemanager/latest/lib/python3.6/site-packages/htmd/molecule/readers.py:778: FutureWarning: Method .as_matrix will be removed in a future version. Use .values instead.
  serials = parsedtopo.serial.as_matrix()
2018-07-06 12:16:50,822 - htmd.molecule.readers - WARNING - Integer values detected in "insertion" field. Your resids might be overflowing. Take care
Projecting trajectories: 100%|██████████████████████████████████████████████████████████████████████████████████████| 2/2 [00:00<00:00,  5.22it/s]
Traceback (most recent call last):
  File "/home/joao/maindisk/SANDBOX/acemanager/latest/lib/python3.6/site-packages/joblib/externals/loky/process_executor.py", line 377, in _process_worker
    call_item = call_queue.get(block=True, timeout=timeout)
  File "/home/joao/maindisk/SANDBOX/acemanager/latest/lib/python3.6/multiprocessing/queues.py", line 113, in get
    return _ForkingPickler.loads(res)
AttributeError: Can't get attribute 'MetricRmsd' on <module 'joblib.externals.loky.backend.popen_loky_posix' from '/home/joao/maindisk/SANDBOX/acemanager/latest/lib/python3.6/site-packages/joblib/externals/loky/backend/popen_loky_posix.py'>
Traceback (most recent call last):
  File "/home/joao/maindisk/SANDBOX/acemanager/latest/lib/python3.6/site-packages/joblib/externals/loky/process_executor.py", line 377, in _process_worker
    call_item = call_queue.get(block=True, timeout=timeout)
  File "/home/joao/maindisk/SANDBOX/acemanager/latest/lib/python3.6/multiprocessing/queues.py", line 113, in get
    return _ForkingPickler.loads(res)
AttributeError: Can't get attribute 'MetricRmsd' on <module 'joblib.externals.loky.backend.popen_loky_posix' from '/home/joao/maindisk/SANDBOX/acemanager/latest/lib/python3.6/site-packages/joblib/externals/loky/backend/popen_loky_posix.py'>
Traceback (most recent call last):
  File "htmd/projections/metricrmsd.py", line 161, in <module>
    data2 = metr2.project()
  File "/home/joao/maindisk/SANDBOX/acemanager/latest/lib/python3.6/site-packages/htmd/projections/metric.py", line 143, in project
    results = aprun(total=numSim, desc='Projecting trajectories')(delayed(_processSim)(self.simulations[i], self.projectionlist, uqMol, self.skip) for i in range(numSim))
  File "/home/joao/maindisk/SANDBOX/acemanager/latest/lib/python3.6/site-packages/htmd/parallelprogress.py", line 18, in <lambda>
    return lambda x: Parallel(**joblib_args)(tqdm_f(x, tq_args))
  File "/home/joao/maindisk/SANDBOX/acemanager/latest/lib/python3.6/site-packages/joblib/parallel.py", line 962, in __call__
    self.retrieve()
  File "/home/joao/maindisk/SANDBOX/acemanager/latest/lib/python3.6/site-packages/joblib/parallel.py", line 865, in retrieve
    self._output.extend(job.get(timeout=self.timeout))
  File "/home/joao/maindisk/SANDBOX/acemanager/latest/lib/python3.6/site-packages/joblib/_parallel_backends.py", line 515, in wrap_future_result
    return future.result(timeout=timeout)
  File "/home/joao/maindisk/SANDBOX/acemanager/latest/lib/python3.6/site-packages/joblib/externals/loky/_base.py", line 431, in result
    return self.__get_result()
  File "/home/joao/maindisk/SANDBOX/acemanager/latest/lib/python3.6/site-packages/joblib/externals/loky/_base.py", line 382, in __get_result
    raise self._exception
joblib.externals.loky.process_executor.BrokenProcessPool: A process in the executor was terminated abruptly while the future was running or pending.
```

For now, our solution will be to pin to 0.11, but let me know if you need any extra info to help solving the issue.

Best,
João

**Comment by lesteve** [comment]
It would be great if you can provide a stand-alone snippet to reproduce the problem.

For reference the original snippet by @TomDLT seems to work fine for joblib 0.12.

**Comment by stefdoerr** [comment]
Minimal example. Seems to break when passing lists of functions or class objects. Works fine with single objects.

```python
from joblib import Parallel, delayed

class MetricBad:
    def __init__(self):
        self.x = 5

    def project(self):
        self.x = 9
        return self.x

def _process(projectionlist):
    for p in projectionlist:
        res = p.project()
    return res

def _process2(projection):
    return projection.project()

results = Parallel(n_jobs=4)(delayed(_process)([MetricBad(), MetricBad()]) for i in range(20))  # Breaks
results = Parallel(n_jobs=4)(delayed(_process2)(MetricBad()) for i in range(20))  # Works

def _process3(funcs):
    for fn in funcs:
        res = fn()
    return res

def return6():
    return 6

def _process4(fn):
    return fn()

results = Parallel(n_jobs=4)(delayed(_process3)([return6, return6]) for i in range(20)) # Breaks also with lists of functions
results = Parallel(n_jobs=4)(delayed(_process4)(return6) for i in range(20)) # Works
```

**Comment by tomMoral** [comment]
This is a known behavior.
Between `0.10` and `0.11`, the default backend of `joblib` have been changed to stop using `fork` to launch new processes. While this solve many issue with third party libraries (such as openMP, openBLAS, openCV,...), it makes the serialization of python object more complex, especially for python function that are part of the `__main__` module.

To cope with this issue, #677 implements a mechanism that uses `cloudpickle` to serialize objects that are detected to be in the `__main__` module. However, the [`_wrap_non_picklable_objects` function](https://github.com/joblib/joblib/pull/677/files#diff-12c1bd53fbe4c0df54e8edee36f2d2e9R193) does not introspect objects. Thus, list of functions from the `__main__` module are not wrapped and cause this failure. For object methods, I am not sure of the why the detection does not work but it is the same principle, `joblib` does not detect that they should be pickled with `cloudpickle` and not regular `pickle`.

The rational behind this detection is that the serialization with `cloudpickle` is slower than with `pickle` so it is better to only use it when needed.

A work around to solve this for your usage would be to wrap the failing function directly using

```python
from joblib import Parallel, delayed
from joblib.parallel import CloudpickledObjectWrapper

class MetricBad:
    def __init__(self):
        self.x = 5

    def project(self):
        self.x = 9
        return self.x

def _process(projectionlist):
    for p in projectionlist:
        res = p.project()
    return res

def _process2(projection):
    return projection.project()

args = [MetricBad(), MetricBad()]
wrapped_args = [CloudpickledObjectWrapper(obj) for obj in args]

results = Parallel(n_jobs=4)(delayed(_process)(wrapped_args)
                             for i in range(20))  # Breaks
results = Parallel(n_jobs=4)(delayed(_process2)(MetricBad())
                             for i in range(20))  # Works

def _process3(funcs):
    for fn in funcs:
        res = fn()
    return res

def return6():
    return 6

def _process4(fn):
    return fn()

wrapped_func = CloudpickledObjectWrapper(return6)

results = Parallel(n_jobs=4)(delayed(_process3)([wrapped_func, wrapped_func])
                             for i in range(20))
results = Parallel(n_jobs=4)(delayed(_process4)(return6) for i in range(20))


```

**Comment by ogrisel** [comment]
@tomMoral Instead of only introspecting the first level arguments to detect whether they are dynamically defined objects, we could also introspect the content of builtin containers such as lists or dicts but that would be very expensive.

An option would be to only inspect the first 10 elements of such containers and assume that the container contents are approximately homogeneous.

For reference, the problem happens at unpickling time in the worker process:

```python-traceback
Traceback (most recent call last):
  File "/home/ogrisel/code/joblib/joblib/externals/loky/process_executor.py", line 381, in _process_worker
    call_item = call_queue.get(block=True, timeout=timeout)
  File "/usr/lib/python3.6/multiprocessing/queues.py", line 113, in get
    return _ForkingPickler.loads(res)
AttributeError: Can't get attribute 'return6' on <module 'joblib.externals.loky.backend.
```

If it had been a pickling issue at pickling time we might have been able to detect it and to automatically fallback to the slow cloudpickle in those cases.

The only way to make pickling of interactively defined functions & classes would be to implement support in the C implementation of the pickle module in the standard library. This is not impossible to do and actually python dev are working on this:

https://lists.gt.net/python/dev/1362786#1362786

However, this won't be available before Python 3.8 in the best case.

**Comment by lesteve** [comment]
I guess another work-around would be to use `backend='multiprocessing'`, is it a work-around we want to recommend though?

**Comment by ogrisel** [comment]
I would rather like to deprecate the multiprocessing backend in the long term.

**Comment by joseph-long** [comment]
Just encountered this with a function that's not defined in `__main__` but rather a `partial()` from `functools`. I seem to remember pickling partials being added in an early version of Python 3.x, so I'm not sure why it's failing to unpickle now...

**Comment by joseph-long** [comment]
Upon further investigation, it only seems to happen when I supply the `n_jobs` argument.

```python
import joblib
from functools import partial

def worker(a, b):
        return a + b

def entry_point(many_a, b):
        do_the_thing = partial(worker, b=b)
        output_files = joblib.Parallel(n_jobs=3)(
            joblib.delayed(do_the_thing)(an_a)
            for an_a in many_a
        )
        print(output_files)

if __name__ == "__main__":
        entry_point([1,2,3], 10)
```

**Comment by CompPhysChris** [comment]
I've seen the same error with respect to passing a partial function.  I'm attaching the output from a notebook that shows the same calculation four times.  One without partial, one with partial but no n_jobs, and one with partial and n_jobs that fails with the function defined in the script/notebook.  It turns out it also doesn't error if the function is imported from another file, even if n_jobs is passed.  This is shown in the forth test.

[joblib partial bug.pdf](https://github.com/joblib/joblib/files/2257776/joblib.partial.bug.pdf)

**Comment by ogrisel** [comment]
@CompPhysChris please do not include the code snippets and error message directly in the body of the github issue (or as a comment) using the triple backtick ` notation: https://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports#use-syntax-highlighting

That makes it easier to copy and paste your snippets to reproduce the problem.

BTW this is fixed in joblib 0.12.2. Feel free to open new issues if you find other similar cases. Don't forget to include the joblib version number in your reports.
